### PR TITLE
When querying workflows during backoff, check if query can wait

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?
Before: if there was not workflow task scheduled just fail.
After: check if query will actually time out while waiting, and if it will not - wait.

## Why?
To reduce avoidable failures. 
Query shouldn't fail if it can wait until successful conditions will be met.

## How did you test it?
Extend existing functional test to cover both situations.

## Potential risks
N/A

## Documentation
N/A

## Is hotfix candidate?
No

### CODE REVIEW

File: service/history/api/queryworkflow/api.go [hunk 1]
```diff
@@ -241,6 +248,27 @@ func Invoke(
...
+		return false, err
+	}
+	startAttr := startEvent.GetWorkflowExecutionStartedEventAttributes()
+	workflowTaskBackoffDuration := timestamp.DurationValue(startAttr.GetFirstWorkflowTaskBackoff())
+
+	workflowStart := mutableState.GetExecutionInfo().StartTime.AsTime().UTC()
+	workflowTaskStart := workflowStart.Add(workflowTaskBackoffDuration)
+
+	deadline, _ := ctx.Deadline()
+	deadline = deadline.UTC()
```
Comment (alexshtin) [inline service/history/api/queryworkflow/api.go:265]: I think you should check 2nd return value (if deadline is set at all).

File: service/history/api/queryworkflow/api.go [hunk 2]
```diff
@@ -241,6 +248,27 @@ func Invoke(
 	}
 }
 
+func queryWillTimeoutsBeforeTaskStart(
```
Comment (alexshtin) [inline service/history/api/queryworkflow/api.go:251]: Let's be more specific about task:
```suggestion
func queryWillTimeoutsBeforeFirstWorkflowTaskStart(
```

File: tests/query_workflow.go [hunk 3]
```diff
@@ -175,27 +174,29 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff() {
...
-	_, err = s.sdkClient.QueryWorkflow(ctx, id, "", testname)
-	s.Error(err)
-	s.ErrorContains(err, consts.ErrWorkflowTaskNotScheduled.Error())
+}
+
+func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff_Pass() {
+	contextTimeout := 10
+	startDelay := 5
+	expectError := false
+	s.queryWorkflow_QueryWhileBackoff(contextTimeout, startDelay, expectError)
```
Comment (alexshtin) [inline tests/query_workflow.go:192]: You should combine them in subtests. See `t.Run()` calls in other tests.

File: tests/query_workflow.go [hunk 4]
```diff
@@ -175,27 +174,29 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff() {
...
-			findBackoffWorkflow = true
-			break
-		}
+	_, err = s.sdkClient.QueryWorkflow(ctx, id, "", testname)
 
-		// wait for the retry, which will have backoff
-		time.Sleep(time.Second)
+	if expectError {
+		s.Error(err)
+		s.ErrorContains(err, consts.ErrWorkflowTaskNotScheduled.Error())
```
Comment (alexshtin) [inline tests/query_workflow.go:181]: Don't refer `const`. Hard code string here. It is a part of contract and should be checked explicitly.

File: tests/query_workflow.go [hunk 5]
```diff
@@ -175,27 +174,29 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff() {
...
 	}
-	s.True(findBackoffWorkflow)
 
-	_, err = s.sdkClient.QueryWorkflow(ctx, id, "", testname)
-	s.Error(err)
-	s.ErrorContains(err, consts.ErrWorkflowTaskNotScheduled.Error())
+}
+
+func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff_Pass() {
+	contextTimeout := 10
```
Comment (alexshtin) [inline tests/query_workflow.go:189]: You need really long context timeout here to make sure it won't fire.

File: tests/query_workflow.go [hunk 6]
```diff
@@ -175,27 +174,29 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff() {
...
+	startDelay := 5
+	expectError := false
+	s.queryWorkflow_QueryWhileBackoff(contextTimeout, startDelay, expectError)
+}
+
+func (s *ClientFunctionalSuite) TestQueryWorkflow_QueryWhileBackoff_Fail() {
+	contextTimeout := 8
+	startDelay := 10
```
Comment (alexshtin) [inline tests/query_workflow.go:197]: Define them as `time.Duration`.

File: service/history/api/queryworkflow/api.go (hunk 1 above)
Comment (ychebotarev) [reply service/history/api/queryworkflow/api.go:265]: done in the next rev.

File: service/history/api/queryworkflow/api.go (hunk 2 above)
Comment (ychebotarev) [reply service/history/api/queryworkflow/api.go:251]: I think it is too verbose, but ok

File: tests/query_workflow.go (hunk 3 above)
Comment (ychebotarev) [reply tests/query_workflow.go:192]: done in a different way in the next rev

File: tests/query_workflow.go (hunk 6 above)
Comment (ychebotarev) [reply tests/query_workflow.go:197]: done in the next rev

File: tests/query_workflow.go (hunk 5 above)
Comment (ychebotarev) [reply tests/query_workflow.go:189]: it is going to be set to 10 anyway

File: tests/query_workflow.go [hunk 7]
```diff
@@ -146,26 +145,72 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_Consistent_PiggybackQuery() {
...
+	testCases := []struct {
+		contextTimeout time.Duration
+		startDelay     time.Duration
+		expectError    bool
+		workflowFn     any
+		testname       string
+	}{
+		{
+			contextTimeout: time.Duration(10) * time.Second,
+			startDelay:     time.Duration(5) * time.Second,
```
Comment (alexshtin) [inline tests/query_workflow.go:174]:
```suggestion
			contextTimeout: 10 * time.Second,
			startDelay:     5 * time.Second,
```

File: tests/query_workflow.go [hunk 8]
```diff
@@ -146,26 +145,72 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_Consistent_PiggybackQuery() {
...
+			contextTimeout: time.Duration(8) * time.Second,
+			startDelay:     time.Duration(10) * time.Second,
+			expectError:    true,
+			workflowFn:     workflowFnPass,
+			testname:       testnamePass,
+		},
+	}
+
+	for _, tc := range testCases {
+		s.testQueryWorkflow_QueryWhileBackoff(
```
Comment (alexshtin) [inline tests/query_workflow.go:189]: Use `t.Run()` here.

File: tests/query_workflow.go [hunk 9]
```diff
@@ -146,26 +145,72 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_Consistent_PiggybackQuery() {
...
+			tc.contextTimeout, tc.startDelay, tc.expectError, tc.workflowFn, tc.testname,
+		)
+	}
+}
+
+func (s *ClientFunctionalSuite) testQueryWorkflow_QueryWhileBackoff(
```
Comment (alexshtin) [inline tests/query_workflow.go:195]: Please inline this function in the test.

File: tests/query_workflow.go [hunk 10]
```diff
@@ -146,26 +145,72 @@ func (s *ClientFunctionalSuite) TestQueryWorkflow_Consistent_PiggybackQuery() {
...
 			return "should-reach-here", nil
 		})
-		return "", temporal.NewApplicationError("retry-me", "test-error")
+		return "", nil
+	}
+	workflowFnPass := func(ctx workflow.Context) (string, error) {
+		_ = workflow.SetQueryHandler(ctx, testnamePass, func() (string, error) {
+			return "should-reach-here", nil
+		})
+		return "", nil
```
Comment (alexshtin) [inline tests/query_workflow.go:162]: Why do you need different functions? Also why different query handler names?

File: tests/query_workflow.go (hunk 8 above)
Comment (ychebotarev) [reply tests/query_workflow.go:189]: I prefer "table" test cases over run, they are cleaner..

File: tests/query_workflow.go (hunk 9 above)
Comment (ychebotarev) [reply tests/query_workflow.go:195]: done in next rev

File: tests/query_workflow.go (hunk 10 above)
Comment (ychebotarev) [reply tests/query_workflow.go:162]: without it engine panics with "function with the same name is already registered".

File: tests/query_workflow.go (hunk 4 above)
Comment (ychebotarev) [reply tests/query_workflow.go:227]: discussed offline. will continue as is.

File: tests/query_workflow.go (hunk 9 above)
Comment (ychebotarev) [reply tests/query_workflow.go]: actually not happening...
doing this trigger linter error (defer inside the loop).
I will revert this change.

**Review by alexshtin — APPROVED** [review]

**Pull request merged**